### PR TITLE
pkg/trace: remove gogo protobuf

### PR DESCRIPTION
**Pull request by knusbaum**
### What does this PR do?

This commit replaces gogo protobuf with the first-party protobuf generator for Go, and regenerates the protocol buffer code.

### Motivation

Gogo is no longer maintained, so we should move off of it. Also, we want to migrate our proto files into the agent's main proto package along with all the other protobuf stuff, which will require us to use the same code generator as the rest of the agent

### Additional Notes

Currently, this code requires a third-party generator in addition to the main protobuf generator.
This is required to add msgpack struct tags to the generated protobuf structs.

github.com/favadi/protoc-go-inject-tag

The commands to generate the code are:
```
protoc -I. --go_out=paths=source_relative:. span.proto tracer_payload.proto agent_payload.proto stats.proto

#go install github.com/favadi/protoc-go-inject-tag@latest
protoc-go-inject-tag -input=span.pb.go
protoc-go-inject-tag -input=tracer_payload.pb.go
protoc-go-inject-tag -input=agent_payload.pb.go
```

These should maybe live in go:generate, however go:generate currently contains msgpack generation commands which clobber our hand-written modifications to the msgpack encoders.

### Possible Drawbacks / Trade-offs

It's not clear if github.com/favadi/protoc-go-inject-tag is going to be ok to use. We need to check with someone from core or platform to make sure we can add tags in this way. Otherwise, we will have to find another way to add these tags.

One other option is the protoc plugin https://github.com/srikrsna/protoc-gen-gotag

### Describe how to test/QA your changes

This change completely rewrites the protobuf types we use both for ingesting msgpack payloads and for sending to intake, so this needs to be reviewed and tested carefully. 

- [x] Try generating the protobuf
- [x] Try running a local application with non-trivial traces and ensure they appear correctly in the backend

In addition to basic local testing, we should monitor deployments to staging to ensure we don't see any disruptions. Dogfooding may catch corner cases we have blind spots for.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/trace/pb/stats.proto
Comment (knusbaum) [inline pkg/trace/pb/stats.proto:4]: The `.pb.go` file associated with this `.proto` file has not been regenerated as part of this PR. I am going to move this `.proto` file change into another PR as it requires extensive changes to helper functions, tests, etc.  Ignore this file for now.

File: pkg/trace/pb/span.proto
Comment (knusbaum) [inline pkg/trace/pb/span.proto:8]: The `@gotags` are here for the 3rd-party tag-adder thing.

**Review by katiehockman — COMMENTED** [review]
Let's wait to hear back from the other teams about using the new third-party tool, but I can at least confirm that local testing worked fine. I re-generated the protos locally and observed that the generated code didn't change, and I ran with non-trivial traces. e.g.
```
span, ctx := tracer.StartSpanFromContext(ctx, "parent",
	tracer.ServiceName("something"),
	tracer.ResourceName(string(b)),
	tracer.Tag(ext.ResourceName, `SELECT * FROM TABLE WHERE id = 54321`),
	tracer.SpanType(ext.SQLType))
```
(which had a child span as well)

**Review by truthbk — APPROVED** [review]
Looks good to me, no issues whatsoever with introducing a new tool for the tag injection.

**Comment by truthbk** [comment]
CI is seriously broken in the PR though, so we should try to get that fixed before merging though.

File: pkg/trace/api/otlp_test.go
Comment (mx-psi) [inline pkg/trace/api/otlp_test.go:549]: I don't understand why this change is needed/how it is related to gogo protobuf, could you elaborate?

Comment (truthbk) [reply pkg/trace/api/otlp_test.go:549]: Because the new protobuf definitions incorporate a mutex; mutexes shouldn't be copied. So either we use pointers, or we make "manual" copies that don't copy the mutex. I've actually done that in a couple of places where copies seemed necessary, but mostly just switched to references instead of copies. In any case I was waiting for @knusbaum review to better understand if the changes are OK.

Comment (truthbk) [reply pkg/trace/api/otlp_test.go:549]: Also, we still have to fix the CircleCI integration tests. Looking into it.

Comment (mx-psi) [reply pkg/trace/api/otlp_test.go:549]: Okay, thanks for the explanation! Maybe you can mark as draft until the test suite is fixed and things are reviewed by @knusbaum? (So that we know when to review)

Comment (truthbk) [reply pkg/trace/api/otlp_test.go:549]: It's fixed.

**Review by mx-psi — APPROVED** [review]
Changes in OTel files LGTM

**Review by katiehockman — APPROVED** [review]

**Review by KSerrania — APPROVED** [review]

**Pull request merged**
